Leadership transfer

i0rek opened the pull request:
This PR is implementing the leadership transfer extension described in the [thesis chap 3.10](https://ramcloud.stanford.edu/~ongaro/thesis.pdf).

Background:

Consul is performing some setup after acquiring leadership. It is possible that the setup fails, but there is no good way to step down as a leader. It is possible to use `DemoteVoter` as show in https://github.com/hashicorp/consul/pull/5247, but this is suboptimal because it relies on Consul's autopilot to promote the old leader to a voter again.
Since there is a perfectly fine way described in the thesis: leadership transfer extension, we decided to implement that instead. Doing it this way also helps other teams, since it is more generic.

The necessary steps to perform are:

1. Leader picks target to transition to
2. Leader stops accepting client requests
3. Leader makes sure to replicate logs to the target
4. Leader sends TimeoutNow RPC request
5. Target receives TimeoutNow request, which triggers an election
6a. If the election is successful, a message with the new term will make the old leader step down
6b. if after electiontimeout the leadership transfer did not complete, the old leader resumes operation

Todo:

* [ ] mention leadership transfer in the readme
* [x] add fuzzy test for leadership transfer
* [x] make the leader semi respond to stuff until transition is either completed or failed
* [x] if after ~electiontimeout the leadership transfer did not complete, the old leader resumes operation
* [x] add leadershiptransfer flag to `RequestVote` as described in [chap 4.2.3 Disruptive servers](https://ramcloud.stanford.edu/~ongaro/thesis.pdf)
* [x] expire `LeaderLeaseTimeout` before leadership transfer as described in [chap 6.4.1 Using clocks to reduce messaging for read-only queries](https://ramcloud.stanford.edu/~ongaro/thesis.pdf)
* [x] are we calling the replication correctly?
* [x] add documentation in the code

Resources:

* https://github.com/etcd-io/etcd/tree/master/raft

i0rek commented:
@mitchellh Absolutely, I added the todos!

banks commented:
@tylertreat FYI. I know you were interested in this!

banks commented on raft.go:
```diff
@@ -652,6 +702,75 @@ func (r *Raft) verifyLeader(v *verifyFuture) {
+	// Step 3: reset leader lease. Since a new leader might be choosen in a bit,
+	// we don't want this leader to have the right to answer read request based on
+	// this lease.
+	r.leaderState.lease = time.After(0)
```
Hmm I'm not sure this is right for a very confusing reason!

The concept of a lease in our leader implementation is not really quite the same as the lease mechanism described in `6.4.1 Using clocks to reduce messaging for read-only queries` of the thesis.

That lease mechanism's only purpose is to support almost-linearizable reads (assuming clocks are enough in sync) without the message round trips. 

Our raft lib doesn't actually have a concept of fully linearizable reads in the same way - we assume that reads are made directly from the FSM implemented on top of this on the leader (not fully linearizable since the leader might have lost an election during partition but not noticed yet because still inside election timeout). Where fully-consistent reads are needed (i.e. `?consistent` in Consul) we still don't process the read through raft log as some raft implementations do, instead we first get a snapshot of current leader FSM state, then run verifyLeader which essentially re-runs leader election to ensure a quorum still support the leader, and then answer the request from the snapshot that we had which is linearizable because you know the server was still the only leader at the point that snapshot was valid.

The lease mechanism in our raft lib, is setup in a similar way to the one in the thesis, but in practice it's a timer that just tells the leader how often to check heartbeats are still coming in.

So if you stop the lease here, then as soon as this method exits below (right after sending TimeoutNow, but _probably before_ the actual new leader takes over, the leaderloop above will activate the `<-lease` case, re-check the heartbeats for a quorum are still coming in, **and then rest the lease timeout again**, in other words this is kind of a no-op - it just causes the heartbeat check to happen a few ms sooner than it might.

Worst case though, if heartbeats have stopped for a while (for example because it took us a few seconds to replicate log entries above) then this will mean that immediately after we send `TimeoutNow` and return here, the leader will hit the `<-lease` case and then find it's lost lease and step down - possibly before the new leader had a chance to win it's election and take over.

Ha ha the thesis in section six says:

>  Problems due to stale reads have already been discovered in two third-party Raft implementations [45]

One of those was this one before we implemented verifyLeader 😄 .

Actually our verifyLeader mechanism technically isn't the same as the one proposed in section 6 either... We don't record the leader's commit index before and wait for it to be applied to the FSM etc. in practice given the bounds on how fast other things can happen I think it's sufficient.

The tl;dr here is that I think we can just ignore this and be no less correct than now - reads that we consider "linearizable" use a different mechanism that perform their own "heartbeats" (in thesis language, actually we call them votes in verifyLeader) anyway so we effectively don't use the `lease` as a way to assume reads are consistent.

i0rek replied:
thank you for that explanation! removed it from the pr.

banks replied:
Might be nice to keep a comment in the code to link to this for when we come back later and wonder why we didn't follow the thesis on that?

i0rek commented on raft.go:
```diff
@@ -652,6 +702,75 @@ func (r *Raft) verifyLeader(v *verifyFuture) {
+			r.logger.Printf("[DEBUG] raft: %v", err.Error())
+			future.respond(err.Error())
+			return
+		}
```
You are correct! It could potentially sit here forever trying to catch up. I will change it so that it is time out.

i0rek commented on raft.go:
```diff
@@ -652,6 +702,75 @@ func (r *Raft) verifyLeader(v *verifyFuture) {
+		err := fmt.Errorf("cannot find replication state for %v, aborting leadership transfer", future)
+		r.logger.Printf("[DEBUG] raft: %v", err)
+		future.respond(err)
+		return
```
You are right we need to add `r.leaderState.leadershipTransferInProgress = false` here. I don't think we can defer that since the happy path needs it to be enabled. Because the target becomes leader after we returned from this function and we cannot accept client requests in the meanwhile.

mkeeler commented on raft.go:
```diff
@@ -505,6 +522,14 @@ func (r *Raft) leaderLoop() {
+			}
+			r.leadershipTransfer(future)
+
 		case <-r.leaderState.commitCh:
```
Looks like you brought up the blocking nature of `leadershipTransfer` and it potentially taking a long time causing other issues in a different comment.

If we make it not block then we definitely should handle this case here.

banks replied:
It partially blocks (replication blocks but then it runs the rest in a go routine and returns) in the implementation I commented on which is both a performance issue and a bug! Not checked the recent changes here yet though which maybe resolved that.

banks replied:
For future reference the reason this is fine is that `commitCh` is used to be notified by replication code when a log has been acked by enough followers to be considered committed. Allowing this still is important in case the leadership transfer doesn't work and we are still leader but dropped the ball on tracking which logs got committed by followers.

i0rek commented on api.go:
```diff
@@ -461,6 +475,7 @@ func NewRaft(conf *Config, fsm FSM, logs LogStore, stable StableStore, snaps Sna
 		configurationsCh:      make(chan *configurationsFuture, 8),
 		bootstrapCh:           make(chan *bootstrapFuture),
 		observers:             make(map[uint64]*Observer),
+		leadershipTransferCh:  make(chan *leadershipTransferFuture, 64),
```
I randomly picked that number. That was before I returned `ErrLeadershipTransferInProgress` instead of waiting to be able to write to the channel. I changed it to use `0` capacity.

i0rek replied:
oh well, of course it needs to be 1. 🤦‍♂️

i0rek commented:
I addressed the fact, that the replication needs to be included in the timeout as well. I think it is much better now, because not only timeouts, but also loss of leadership can stop the transfer at any point in time: https://github.com/hashicorp/raft/blob/ded6ce1d88c0dba7140992d16eb9c138f09ba38d/raft.go#L540-L558.

i0rek commented:
I also defer resetting the flag for leadershiptransfer in progress now: https://github.com/hashicorp/raft/blob/ded6ce1d88c0dba7140992d16eb9c138f09ba38d/raft.go#L735-L736.

i0rek commented on api.go:
```diff
@@ -1006,3 +1021,34 @@ func (r *Raft) LastIndex() uint64 {
+
+// LeadershipTransferToServer does the same as LeadershipTransfer but takes a
+// server in the arguments in case a leadership should be transitioned to a
+// specific server in the cluster.
```
It will update the server and if it happens to take longer than `electionTimeout` it will abort the transfer.

i0rek commented on raft.go:
```diff
@@ -652,6 +702,75 @@ func (r *Raft) verifyLeader(v *verifyFuture) {
+		case err := <-verify.errCh:
+			r.logger.Printf("[DEBUG] raft: resetting leadership transfer: %s", err)
+			future.respond(nil)
+			r.leaderState.leadershipTransferInProgress = false
```
I put this into a defer: https://github.com/hashicorp/raft/blob/c2eb642215d568b9925f86781a9b629c5fd1457a/raft.go#L755-L759

But since it is still being read from the mainloop, I also put it into an atomic int.

i0rek commented on raft.go:
```diff
@@ -1468,3 +1594,55 @@ func (r *Raft) setState(state RaftState) {
+}
+
+// pickServer returns the first server that is not itself.
+func (r *Raft) pickServer() *Server {
```
I changed the implementation to do exactly that.

i0rek commented:
ok, I am at a point where I addressed the feedback except these two points: 

1) **protocol version:** there is no point checking for version 3 - because not all v3 raft has leadership transfer. should we create v4 for it?
2) **which actions need to be blocked?** The reason why I did it this way is the paper says `1. The prior leader stops accepting new client requests.` which I interpreted as every request that comes from the API. Since only the leader can transfer leadership, and the leader initiates every action, I think what I did is enough. I might be wrong.

i0rek commented:
In a call with @banks we decided, that we don't cut a new version. I will make it clear in the docs that you have to run the latest code in order to use leadership transfer - that v3 is not sufficient.

With regards to blocking actions coming through `commitCh`, I don't think we should do that. The commitCh receives a message when an entry is commited: accepted by the majority of the followers. This doesn't change the raft state. But it changes the state of the leader. However even when we chose to not process these messages, we are basically missing stuff since we are still the leader. If we happen to transfer to a follower that has this commit, everything is fine. If we transfer to a follower that doesn't, it won't win the election. Not processing commitCh doesn't change that.

i0rek commented:
If a leader attempts to transfer leadership to a follower that doesn't have that feature, it will fail gracefully on the follower and the leader. In that case the old leader keeps being the leader:

```
    2019/02/14 00:17:16 [ERR] raft-net: Failed to decode incoming command: unknown rpc type 3
```

and on the leader as well:

```
    2019/02/14 00:17:16 [WARN] consul: failed to transfer leadership: failed to make TimeoutNow RPC to dd9a2744-96ec-78fc-f69c-d7376e207946: EOF
```

i0rek commented on raft.go:
```diff
@@ -505,6 +541,43 @@ func (r *Raft) leaderLoop() {
+			doneCh := make(chan error, 1)
+
+			go func() {
+				select {
```
also read from `shutdownCh` to make sure we don't hang around longer than we need to.

i0rek replied:
This is accounted for because we have

```
defer func() { close(leftLeaderLoop) }()
```

which will close leftLeaderLoop and trigger the case.

mkeeler commented on api.go:
```diff
@@ -1006,3 +1021,34 @@ func (r *Raft) LastIndex() uint64 {
+// is unlikely that no other server is starting an election, it is very likely
+// that the target server is able to win the election.
+func (r *Raft) LeadershipTransfer() Future {
+	if r.protocolVersion < 3 {
```
For the sake of completeness, the answer to my question is that we don't currently support auto-negotiation of Raft protocol versions. So its not so simple to upgrade existing clusters.

banks commented on raft.go:
```diff
@@ -77,12 +77,14 @@ type commitTuple struct {
+	replState                    map[ServerID]*followerReplication
+	notify                       map[*verifyFuture]struct{}
+	stepDown                     chan struct{}
+	lease                        <-chan time.Time
```
Do we need to move this any more? I think you did it because originally you were waiting on it in the transfer until I pointed out that wasn't necessary.

I don't think it matters much so happy to leave it if that's easier but not sure it's necessary to pull  it out of the leader loop method any more?

banks commented on raft.go:
```diff
@@ -341,6 +356,8 @@ func (r *Raft) runLeader() {
 	r.leaderState.replState = make(map[ServerID]*followerReplication)
 	r.leaderState.notify = make(map[*verifyFuture]struct{})
 	r.leaderState.stepDown = make(chan struct{}, 1)
+	r.leaderState.lease = time.After(r.conf.LeaderLeaseTimeout)
```
Hmm is this right? It seems to be e remnant of the original lease stuff that turned out not to be relevant?

banks commented on raft_test.go:
```diff
@@ -2395,6 +2406,275 @@ func TestRaft_ProtocolVersion_Upgrade_2_3(t *testing.T) {
+
+func TestRaft_LeadershipTransferResetsLeaderLease(t *testing.T) {
+	t.Skip("How do I test this?")
+}
```
Shouldn't even be a test any more with lease stuff removed?

i0rek replied:
agreed!

banks commented on raft_test.go:
```diff
@@ -2395,6 +2406,275 @@ func TestRaft_ProtocolVersion_Upgrade_2_3(t *testing.T) {
+
+func TestRaft_LeadershipTransferToUnresponsiveServer(t *testing.T) {
+	t.Skip("How do I test this?")
+}
```
Should we just remove these if we have no way to test them?

i0rek replied:
agreed!

banks commented on raft.go:
```diff
@@ -1468,3 +1632,71 @@ func (r *Raft) setState(state RaftState) {
+		// return early if this server is up to date
+		if state.nextIndex > target {
+			return &server
+		}
```
Is this necessary?

The loop isn't long and would always find the most up-to-date anyway so I don't think it's a big performance win. In some cases it might even work out to be less-than-optimal for transfer time because there may have been 10 new logs recorded since our `getLastIndex` above and the first server is all 10 behind while another server in the list might be only 1 behind.

i0rek replied:
Agreed. Fixed in https://github.com/hashicorp/raft/pull/306/commits/8deca55fc7984ab0a0dd85b9ab29bccdc6711b6e.

banks commented on raft.go:
```diff
@@ -1468,3 +1632,71 @@ func (r *Raft) setState(state RaftState) {
+		if server.ID == r.localID {
+			continue
+		}
+		state, ok := r.leaderState.replState[server.ID]
```
Is this OK? In general leaderState is expected to only be accessed by the single goroutine running the leader loop. Does this always run there? I think it does but worth checking. If so I suggest we update the comments on these new methods that assume they are part of the leader loop to make that explicit?

i0rek replied:
Great catch, it wasn't called from the leaderloop! Plus I found another place where `leaderstate` is accessed outside of the `leaderloop`. Fixed both things.

i0rek replied:
https://github.com/hashicorp/raft/pull/306/commits/20a87dbeaa9522f3e5305dc1fe2c705a8ed22fbb

james-lawrence commented:
> I should also ask - the failing CI tests, are those same tests failing in master and/or passing locally?

@banks, there are a bunch of minor fixes (by freeekanayaka) in PR form already that improve the CI stability of master. might want to consider merging those and rebasing on top.

banks commented on raft.go:
```diff
@@ -1484,3 +1664,66 @@ func (r *Raft) setState(state RaftState) {
+	return nil
+}
+
+// pickServer returns the follower that is most up to date.
```
Should we comment this is only to be called in leader loop?

i0rek replied:
done!

banks commented on raft.go:
```diff
@@ -506,6 +543,74 @@ func (r *Raft) leaderLoop() {
+			}()
+
+			// leaderState.replState is accessed here before
+			// starting leadership asynchronously transfer because
```
```suggestion
			// starting leadership transfer asynchronously because
```

banks commented on raft.go:
```diff
@@ -665,6 +790,54 @@ func (r *Raft) verifyLeader(v *verifyFuture) {
+	r.setLeadershipTransferInProgress(true)
+	defer func() { r.setLeadershipTransferInProgress(false) }()
+
+	for repl.nextIndex <= r.getLastIndex() {
```
This access of the `followerReplication.nextIndex` made me wonder about thread safety. Presumably this is racey since the replication thread is still running and could be writing to the `nextIndex` concurrently?

i0rek replied:
Using `atomic` now to protect from the race. Great catch!
